ffi/memorymanager.cpp

```diff
@@ -0,0 +1,347 @@
+            return (uint8_t *)Addr;
+        }
+    }
+
+    assert(false && "All memory must be pre-allocated");
+}
+
+static uintptr_t requiredPageSize(uintptr_t Size, uint32_t Alignment) {
+    static const size_t PageSize = sys::Process::getPageSizeEstimate();
+    // Use the same calculation as allocateSection because we need to be able to
+    // satisfy it.
+    uintptr_t RequiredSize =
+        Alignment * ((Size + Alignment - 1) / Alignment + 1);
+    // Round up to the nearest page size. Blocks must be page-aligned.
+    return PageSize * ((RequiredSize + PageSize - 1) / PageSize);
+}
+
+void LlvmliteMemoryManager::reserveAllocationSpace(
+    uintptr_t CodeSize, uint32_t CodeAlign, uintptr_t RODataSize,
+    uint32_t RODataAlign, uintptr_t RWDataSize, uint32_t RWDataAlign) {
+    MemoryGroup &MemGroup = CodeMem;
+
+    // MemoryManager does not have functions for releasing memory after it's
+    // allocated. Normally it tries to use any excess blocks that were
+    // allocated due to page alignment, but if we have insufficient free memory
+    // for the request this can lead to allocating disparate memory that can
+    // violate the ARM ABI. Clear free memory so only the new allocations are
+    // used, but do not release allocated memory as it may still be in-use.
+    CodeMem.FreeMem.clear();
+    RODataMem.FreeMem.clear();
+    RWDataMem.FreeMem.clear();
+
+    LLVM_DEBUG(dbgs() << "Code size / align: " << format_hex(CodeSize, 2, true)
+                      << " / " << CodeAlign << "\n");
+    LLVM_DEBUG(dbgs() << "ROData size / align: "
+                      << format_hex(RODataSize, 2, true) << " / " << RODataAlign
+                      << "\n");
+    LLVM_DEBUG(dbgs() << "RWData size / align: "
+                      << format_hex(RWDataSize, 2, true) << " / " << RWDataAlign
+                      << "\n");
+
+    CodeAlign = CodeAlign ? CodeAlign : 16;
+    RODataAlign = RODataAlign ? RODataAlign : 16;
+    RWDataAlign = RWDataAlign ? RWDataAlign : 16;
+    uintptr_t RequiredCodeSize = requiredPageSize(CodeSize, CodeAlign);
+    uintptr_t RequiredRODataSize = requiredPageSize(RODataSize, RODataAlign);
+    uintptr_t RequiredRWDataSize = requiredPageSize(RWDataSize, RWDataAlign);
+    uint64_t TotalSize =
+        RequiredCodeSize + RequiredRODataSize + RequiredRWDataSize;
+
+    LLVM_DEBUG(dbgs() << "Reserving " << format_hex(TotalSize, 2, true)
+                      << " bytes\n");
+
+    // FIXME: It would be useful to define a default allocation size (or add
+    // it as a constructor parameter) to minimize the number of allocations.
+    //
+    // FIXME: Initialize the Near member for each memory group to avoid
+    // interleaving.
+    std::error_code ec;
+    sys::MemoryBlock MB = MMapper.allocateMappedMemory(
+        AllocationPurpose::Code, TotalSize, &MemGroup.Near,
+        sys::Memory::MF_READ | sys::Memory::MF_WRITE, ec);
+    if (ec) {
+        // FIXME: Add error propagation to the interface.
```

Comment:
While we are at it, perhaps returning a bool here isn't too much work and churn.

Reply:
We still need to conform to the interface for `reserveAllocationSpace()` though, so I don't think this is something we can change downstream of LLVM.

Reply:
Oh ok, fair enough.